// @backstage/plugin-events-backend/dist/service/EventsPlugin.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-events-node/alpha');
var pluginEventsNode = require('@backstage/plugin-events-node');
var Router = require('express-promise-router');
var HttpPostIngressEventPublisher = require('./http/HttpPostIngressEventPublisher.cjs.js');
var createEventBusRouter = require('./hub/createEventBusRouter.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var Router__default = /*#__PURE__*/_interopDefaultCompat(Router);

class EventsExtensionPointImpl {
  #httpPostIngresses = [];
  #httpBodyParsers = [];
  setEventBroker(_) {
    throw new Error(
      "setEventBroker is not supported anymore; use eventsServiceRef instead"
    );
  }
  addPublishers(_) {
    throw new Error(
      "addPublishers is not supported anymore; use EventsService instead"
    );
  }
  addSubscribers(_) {
    throw new Error(
      "addSubscribers is not supported anymore; use EventsService instead"
    );
  }
  addHttpPostIngress(options) {
    this.#httpPostIngresses.push(options);
  }
  addHttpPostBodyParser(options) {
    this.#httpBodyParsers.push(options);
  }
  get httpPostIngresses() {
    return this.#httpPostIngresses;
  }
  get httpBodyParsers() {
    return this.#httpBodyParsers;
  }
}
const eventsPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "events",
  register(env) {
    const extensionPoint = new EventsExtensionPointImpl();
    env.registerExtensionPoint(alpha.eventsExtensionPoint, extensionPoint);
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        events: pluginEventsNode.eventsServiceRef,
        database: backendPluginApi.coreServices.database,
        httpAuth: backendPluginApi.coreServices.httpAuth,
        httpRouter: backendPluginApi.coreServices.httpRouter,
        lifecycle: backendPluginApi.coreServices.lifecycle,
        logger: backendPluginApi.coreServices.logger,
        scheduler: backendPluginApi.coreServices.scheduler
      },
      async init({
        config,
        events,
        database,
        httpAuth,
        httpRouter,
        lifecycle,
        logger,
        scheduler
      }) {
        const ingresses = Object.fromEntries(
          extensionPoint.httpPostIngresses.map((ingress) => [
            ingress.topic,
            ingress
          ])
        );
        const bodyParsers = Object.fromEntries(
          extensionPoint.httpBodyParsers.map((option) => [
            option.contentType,
            option.parser
          ])
        );
        const http = HttpPostIngressEventPublisher.HttpPostIngressEventPublisher.fromConfig({
          config,
          events,
          ingresses,
          bodyParsers,
          logger
        });
        const eventsRouter = Router__default.default();
        http.bind(eventsRouter);
        httpRouter.use(eventsRouter);
        const notifyTimeoutMs = config.getOptionalNumber(
          "events.notifyTimeoutMs"
        );
        httpRouter.use(
          await createEventBusRouter.createEventBusRouter({
            database,
            lifecycle,
            logger,
            httpAuth,
            scheduler,
            notifyTimeoutMs
          })
        );
        httpRouter.addAuthPolicy({
          allow: "unauthenticated",
          path: "/http"
        });
      }
    });
  }
});

exports.eventsPlugin = eventsPlugin;
