// @backstage-community/plugin-github-actions/dist/alpha/entityCards.esm.js
import { jsx } from 'react/jsx-runtime';
import { EntityCardBlueprint } from '@backstage/plugin-catalog-react/alpha';
import { isGithubActionsAvailable } from '../components/Router.esm.js';

const entityLatestGithubActionRunCard = EntityCardBlueprint.makeWithOverrides({
  name: "latest-workflow-run",
  disabled: true,
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().optional()
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.LatestWorkflowRunCard, { ...config.props }))
    });
  }
});
const entityLatestGithubActionsForBranchCard = EntityCardBlueprint.makeWithOverrides({
  name: "latest-branch-workflow-runs",
  disabled: true,
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().optional()
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.LatestWorkflowsForBranchCard, { ...config.props }))
    });
  }
});
const entityRecentGithubActionsRunsCard = EntityCardBlueprint.makeWithOverrides({
  name: "recent-workflow-runs",
  config: {
    schema: {
      props: (z) => z.object({
        branch: z.string().optional(),
        dense: z.boolean().default(false),
        limit: z.number().default(5).optional()
      }).default({})
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      filter: isGithubActionsAvailable,
      loader: async () => import('../components/Cards/index.esm.js').then((m) => /* @__PURE__ */ jsx(m.RecentWorkflowRunsCard, { ...config.props }))
    });
  }
});

export { entityLatestGithubActionRunCard, entityLatestGithubActionsForBranchCard, entityRecentGithubActionsRunsCard };
