// @backstage/plugin-catalog-backend-module-github/dist/analyzers/GithubLocationAnalyzer.cjs.js
'use strict';

var integration = require('@backstage/integration');
var rest = require('@octokit/rest');
var lodash = require('lodash');
var parseGitUrl = require('git-url-parse');
var path = require('path');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var parseGitUrl__default = /*#__PURE__*/_interopDefaultCompat(parseGitUrl);

class GithubLocationAnalyzer {
  catalogClient;
  githubCredentialsProvider;
  integrations;
  auth;
  constructor(options) {
    this.catalogClient = options.catalog;
    this.integrations = integration.ScmIntegrations.fromConfig(options.config);
    this.githubCredentialsProvider = options.githubCredentialsProvider || integration.DefaultGithubCredentialsProvider.fromIntegrations(this.integrations);
    this.auth = options.auth;
  }
  supports(url) {
    const integration = this.integrations.byUrl(url);
    return integration?.type === "github";
  }
  async analyze(options) {
    const { url, catalogFilename } = options;
    const { owner, name: repo } = parseGitUrl__default.default(url);
    const catalogFile = catalogFilename || "catalog-info.yaml";
    const extension = path.extname(catalogFile);
    const extensionQuery = !lodash.isEmpty(extension) ? `extension:${extension.replace(".", "")}` : "";
    const query = `filename:${catalogFile} ${extensionQuery} repo:${owner}/${repo}`;
    const integration = this.integrations.github.byUrl(url);
    if (!integration) {
      throw new Error("Make sure you have a GitHub integration configured");
    }
    const { token: githubToken } = await this.githubCredentialsProvider.getCredentials({
      url
    });
    const octokitClient = new rest.Octokit({
      auth: githubToken,
      baseUrl: integration.config.apiBaseUrl
    });
    const searchResult = await octokitClient.search.code({ q: query }).catch((e) => {
      throw new Error(`Couldn't search repository for metadata file, ${e}`);
    });
    const exists = searchResult.data.total_count > 0;
    if (exists) {
      const repoInformation = await octokitClient.repos.get({ owner, repo }).catch((e) => {
        throw new Error(`Couldn't fetch repo data, ${e}`);
      });
      const defaultBranch = repoInformation.data.default_branch;
      const result = await Promise.all(
        searchResult.data.items.map((i) => `${lodash.trimEnd(url, "/")}/blob/${defaultBranch}/${i.path}`).map(async (target) => {
          const addLocationResult = await this.catalogClient.addLocation(
            {
              type: "url",
              target,
              dryRun: true
            },
            { credentials: await this.auth.getOwnServiceCredentials() }
          );
          return addLocationResult.entities.map((e) => ({
            location: { type: "url", target },
            isRegistered: !!addLocationResult.exists,
            entity: e
          }));
        })
      );
      return { existing: result.flat() };
    }
    return { existing: [] };
  }
}

exports.GithubLocationAnalyzer = GithubLocationAnalyzer;
